Dynamic roles improvements

sureshanaparti opened the pull request:
## Description
<!--- Describe your changes in detail -->

This PR addresses the below improvements for current Dynamic roles functionality.
- Create a role from any of the existing role, using new parameter roleid in createRole API.
- Import a role with its rules, using a new importRole API.
- New default roles for Read-Only and Support Admin & User.
- No modifications (update or delete) allowed for Default roles.

- Cleanup of old NetApp APIs from role_permissions table.

UI changes PR: https://github.com/apache/cloudstack-primate/pull/353

<!-- For new features, provide link to FS, dev ML discussion etc. -->
<!-- In case of bug fix, the expected and actual behaviours, steps to reproduce. -->

<!-- When "Fixes: #<id>" is specified, the issue/PR will automatically be closed when this PR gets merged -->
<!-- For addressing multiple issues/PRs, use multiple "Fixes: #<id>" -->
<!-- Fixes: # -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [x] Cleanup (Code refactoring and cleanup, that may add test cases)

## Screenshots (if appropriate):

## How Has This Been Tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->
**Manual Tests using API**

- Create a role using existing role.

http://myms:8080/client/api?command=createRole&response=json&name=Test%20CloneUser01&description=Test%20CloneUser01&roleid=ca9871c2-8ea7-11ea-944e-c2865825b006

{"createroleresponse":{"role":{"id":"c3888c27-66b6-4a39-96ee-9883fef326ba","name":"Test CloneUser01","type":"User","description":"Test CloneUser01"}}}

- Create a role using both role type and role id.

http://myms:8080/client/api?command=createRole&response=json&type=Admin&name=Test%20CloneResourceAdmin&description=Test%20CloneResourceAdmin&roleid=ca984d00-8ea7-11ea-944e-c2865825b006

{"createroleresponse":{"uuidList":[],"errorcode":431,"cserrorcode":9999,"errortext":"Both role type and role ID should not be specified"}}

- Create a role without role type and role id.

http://myms:8080/client/api?command=createRole&response=json&name=Test%20CloneResourceAdmin&description=Test%20CloneResourceAdmin

{"createroleresponse":{"uuidList":[],"errorcode":431,"cserrorcode":9999,"errortext":"Neither role type nor role ID is provided"}}

- Import role

http://myms:8080/client/api?command=importRole&response=json&name=ImportTestUser&type=User&description=Test%20Import%20Role&rules[0].rule=create*&rules[0].permission=allow&rules[0].description=create%20rule&rules[1].rule=list*&rules[1].permission=allow&rules[1].description=listing&forced=true

{"importroleresponse":{"role":{"id":"8306f86d-ab33-48b8-8263-535a0da5be04","name":"ImportTestUser","type":"User","description":"Test Import Role"}}}

- Import role with same name without force

http://myms:8080/client/api?command=importRole&response=json&name=ImportTestUser&type=User&description=Test%20Import1%20Role&rules[0].rule=create*&rules[0].permission=allow&rules[0].description=create%20rule&rules[1].rule=list*&rules[1].permission=allow&rules[1].description=listings

{"importroleresponse":{"uuidList":[],"errorcode":530,"cserrorcode":4250,"errortext":"Role already exists"}}

**Unit Tests**

[INFO] Running org.apache.cloudstack.api.command.test.ImportRoleCmdTest
[INFO] Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.032 s - in org.apache.cloudstack.api.command.test.ImportRoleCmdTest
[INFO] Running org.apache.cloudstack.api.command.test.CreateRoleCmdTest
[INFO] Tests run: 5, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.018 s - in org.apache.cloudstack.api.command.test.CreateRoleCmdTest

**Smoke Tests**

=== TestName: test_default_role_deletion | Status : SUCCESS ===

=== TestName: test_role_account_acls | Status : SUCCESS ===

=== TestName: test_role_account_acls_multiple_mgmt_servers | Status : SUCCESS ===

=== TestName: test_role_inuse_deletion | Status : SUCCESS ===

_**=== TestName: test_role_lifecycle_clone | Status : SUCCESS ===**_

=== TestName: test_role_lifecycle_create | Status : SUCCESS ===

=== TestName: test_role_lifecycle_delete | Status : SUCCESS ===

_**=== TestName: test_role_lifecycle_import | Status : SUCCESS ===**_

=== TestName: test_role_lifecycle_list | Status : SUCCESS ===

=== TestName: test_role_lifecycle_update | Status : SUCCESS ===

=== TestName: test_role_lifecycle_update_role_inuse | Status : SUCCESS ===

=== TestName: test_rolepermission_lifecycle_concurrent_updates | Status : SUCCESS ===

=== TestName: test_rolepermission_lifecycle_create | Status : SUCCESS ===

=== TestName: test_rolepermission_lifecycle_delete | Status : SUCCESS ===

=== TestName: test_rolepermission_lifecycle_list | Status : SUCCESS ===

=== TestName: test_rolepermission_lifecycle_update | Status : SUCCESS ===

=== TestName: test_rolepermission_lifecycle_update_permission | Status : SUCCESS ===

=== TestName: test_rolepermission_lifecycle_update_permission_negative | Status : SUCCESS ===
<!-- Please read the [CONTRIBUTING](https://github.com/apache/cloudstack/blob/master/CONTRIBUTING.md) document -->

blueorangutan commented:
Packaging result: ✔centos7 ✔debian. JID-1222

sureshanaparti commented:
@rhtyd Please review the changes. Thanks.

blueorangutan commented:
Packaging result: ✔centos7 ✔debian. JID-1223

rhtyd commented:
@blueorangutan test

blueorangutan commented:
@rhtyd a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

rhtyd commented on api/src/main/java/org/apache/cloudstack/api/command/admin/acl/ImportRoleCmd.java:
```diff
@@ -0,0 +1,139 @@
+import com.cloud.user.Account;
+import com.google.common.base.Strings;
+
+@APICommand(name = ImportRoleCmd.APINAME, description = "Imports a role", responseObject = RoleResponse.class,
```
nit - Make description more verbose, something like 'Imports a role based on provided map of rule permissions...'

sureshanaparti replied:
> nit - Make description more verbose, something like 'Imports a role based on provided map of rule permissions...'

Updated

rhtyd commented on api/src/main/java/org/apache/cloudstack/api/command/admin/acl/ImportRoleCmd.java:
```diff
@@ -0,0 +1,139 @@
+    private Map rules;
+
+    @Parameter(name = ApiConstants.FORCED, type = CommandType.BOOLEAN,
+            description = "Force create a role")
```
nit - fix description to say what happens in case of a role name matches (i.e. explain for the users that it will override/replace the previosly created role with the provided parameters).

sureshanaparti replied:
> nit - fix description to say what happens in case of a role name matches (i.e. explain for the users that it will override/replace the previosly created role with the provided parameters).

Updated

rhtyd commented on api/src/main/java/org/apache/cloudstack/api/command/admin/acl/RoleCmd.java:
```diff
@@ -18,11 +18,41 @@
+    /////////////////////////////////////////////////////
+
+    @Parameter(name = ApiConstants.TYPE, type = CommandType.STRING, description = "The type of the role, valid options are: Admin, ResourceAdmin, DomainAdmin, User")
+    private String roleType;
```
I get the reuse with the createRole API, but do we want these parameters for the updateRole API?

sureshanaparti replied:
> I get the reuse with the createRole API, but do we want these parameters for the updateRole API?

yep, UpdateRoleCmd also inherits the same RoleCmd.

rhtyd commented on api/src/main/java/org/apache/cloudstack/api/response/RoleResponse.java:
```diff
@@ -43,6 +43,10 @@
     private String roleDescription;
 
+    @SerializedName(ApiConstants.IS_DEFAULT)
+    @Param(description = "true if role is default, false otherwise")
```
nit - can you also explain what it means for a role to be default (default could be misleading, perhaps what you mean is if the role is built-in).

sureshanaparti replied:
default meaning this role(s) comes by default with cloudstack installation / upgrade. "default" used, in sync with similar usage at the offerings, accounts in CS.

rhtyd commented on engine/schema/src/main/resources/META-INF/db/schema-41310to41400-cleanup.sql:
```diff
@@ -26,3 +26,6 @@ UPDATE `cloud`.`storage_pool` SET path="/var/lib/libvirt/images" WHERE path="/va
 ALTER TABLE `region` DROP INDEX `id_3`;
+
+-- remove the old NetApp storage APIs (unsupported since 4.12) from role_permissions
+DELETE from `cloud`.`role_permissions` WHERE rule IN ('createPool', 'modifyPool', 'deletePool', 'listPools', 'associateLun', 'dissociateLun', 'createLunOnFiler', 'destroyLunOnFiler', 'listLunsOnFiler', 'createVolumeOnFiler', 'destroyVolumeOnFiler', 'listVolumesOnFiler');
```
We want the DB changes in the 4.14 -> 4.15 upgrade path.

sureshanaparti replied:
> We want the DB changes in the 4.14 -> 4.15 upgrade path.

Moved to 4.15 upgrade path

rhtyd commented on engine/schema/src/main/resources/META-INF/db/schema-41310to41400.sql:
```diff
@@ -379,3 +379,81 @@ CREATE TABLE IF NOT EXISTS `cloud`.`kubernetes_cluster_details` (
     PRIMARY KEY(`id`),
     CONSTRAINT `fk_kubernetes_cluster_details__cluster_id` FOREIGN KEY `fk_kubernetes_cluster_details__cluster_id`(`cluster_id`) REFERENCES `kubernetes_cluster`(`id`) ON DELETE CASCADE
 ) ENGINE=InnoDB DEFAULT CHARSET=utf8;
+
```
We want the DB changes in the 4.14 -> 4.15 upgrade path.

sureshanaparti replied:
> We want the DB changes in the 4.14 -> 4.15 upgrade path.

Moved to 4.15 upgrade path

rhtyd commented on engine/schema/src/main/resources/META-INF/db/schema-41310to41400.sql:
```diff
@@ -379,3 +379,81 @@ CREATE TABLE IF NOT EXISTS `cloud`.`kubernetes_cluster_details` (
 ) ENGINE=InnoDB DEFAULT CHARSET=utf8;
+
+ALTER TABLE `cloud`.`roles` ADD COLUMN `is_default` tinyint(1) NOT NULL DEFAULT '0' COMMENT 'is this a default role';
+UPDATE `cloud`.`roles` SET `is_default` = 1 WHERE id IN (1, 2, 3, 4);
```
Please check in code if there exists any restriction on ID for default/built-in roles.

sureshanaparti replied:
> Please check in code if there exists any restriction on ID for default/built-in roles.

The role ids (1,2,3,4) are hard-coded when upgraded to dynamic roles access checker. The same roles are now marked as default roles in cloudstack.

rhtyd commented on engine/schema/src/main/resources/META-INF/db/schema-41310to41400.sql:
```diff
@@ -379,3 +379,81 @@ CREATE TABLE IF NOT EXISTS `cloud`.`kubernetes_cluster_details` (
+UPDATE `cloud`.`roles` SET `is_default` = 1 WHERE id IN (1, 2, 3, 4);
+
+-- Updated Default CloudStack roles with read-only and support admin and user roles
+INSERT INTO `cloud`.`roles` (`uuid`, `name`, `role_type`, `description`, `is_default`) VALUES (UUID(), 'Read-Only Admin', 'Admin', 'Default read-only admin role', 1) ON DUPLICATE KEY UPDATE name=name;
```
Why not change the name to reflect these are default or `built-in` roles, should we be renaming any existing roles matching this? It could lead to side-effects, I would've preferred failing an upgrade instead of silent override.

sureshanaparti replied:
this is in sync with existing default roles with id (1,2,3,4). Shall I prepend all these role names with "Default" ?

rhtyd commented on engine/schema/src/main/resources/META-INF/db/schema-41310to41400.sql:
```diff
@@ -379,3 +379,81 @@ CREATE TABLE IF NOT EXISTS `cloud`.`kubernetes_cluster_details` (
+-- Role permissions for Read-Only User
+SELECT id INTO @ReadOnlyUserRoleId FROM `cloud`.`roles` WHERE name = 'Read-Only User' AND is_default = 1;
+SELECT @ReadOnlyUserSortOrder:=-1;
+INSERT INTO `cloud`.`role_permissions` (`uuid`, `role_id`, `rule`, `permission`, `sort_order`) SELECT UUID(), @ReadOnlyUserRoleId, rule, 'ALLOW', @ReadOnlyUserSortOrder:=@ReadOnlyUserSortOrder+1 FROM `cloud`.`role_permissions` WHERE role_id = 4 AND permission = 'ALLOW' AND rule LIKE 'list%';
```
Should the `rule` here be `list*` or `get*` etc.? Or was `rule` declared somewhere I missed ?

sureshanaparti replied:
> Should the `rule` here be `list*` or `get*` etc.? Or was `rule` declared somewhere I missed ?

the 'list*' and 'get*' rules for the User role are added for read-only user

rhtyd commented on engine/schema/src/main/resources/META-INF/db/schema-41310to41400.sql:
```diff
@@ -379,3 +379,81 @@ CREATE TABLE IF NOT EXISTS `cloud`.`kubernetes_cluster_details` (
+-- Role permissions for Admin-Support
+SELECT id INTO @AdminSupportRoleId FROM `cloud`.`roles` WHERE name = 'Admin-Support' AND is_default = 1;
+SELECT @AdminSupportSortOrder:=-1;
+INSERT INTO `cloud`.`role_permissions` (`uuid`, `role_id`, `rule`, `permission`, `sort_order`) SELECT UUID(), @AdminSupportRoleId, rule, 'ALLOW', @AdminSupportSortOrder:=@AdminSupportSortOrder+1 FROM `cloud`.`role_permissions` WHERE role_id = @ReadOnlyAdminRoleId AND permission = 'ALLOW';
```
Same as above about the `rule` use?

sureshanaparti replied:
> Same as above about the `rule` use?

rules for the Admin support role are added from read-only admin with additional api rules

rhtyd commented on engine/schema/src/main/resources/META-INF/db/schema-41310to41400.sql:
```diff
@@ -379,3 +379,81 @@ CREATE TABLE IF NOT EXISTS `cloud`.`kubernetes_cluster_details` (
+-- Role permissions for User-Support
+SELECT id INTO @UserSupportRoleId FROM `cloud`.`roles` WHERE name = 'User-Support' AND is_default = 1;
+SELECT @UserSupportSortOrder:=-1;
+INSERT INTO `cloud`.`role_permissions` (`uuid`, `role_id`, `rule`, `permission`, `sort_order`) SELECT UUID(), @UserSupportRoleId, rule, 'ALLOW', @UserSupportSortOrder:=@UserSupportSortOrder+1 FROM `cloud`.`role_permissions` WHERE role_id = @ReadOnlyUserRoleId AND permission = 'ALLOW';
```
Same as above about the `rule` use?

sureshanaparti replied:
> Same as above about the `rule` use?

rules for the User support role are added from read-only user with additional api rules

rhtyd commented on server/src/main/java/org/apache/cloudstack/acl/RoleManagerImpl.java:
```diff
@@ -146,10 +150,89 @@ public RoleVO doInTransaction(TransactionStatus status) {
+            public RoleVO doInTransaction(TransactionStatus status) {
+                RoleVO newRoleVO = roleDao.persist(new RoleVO(name, role.getRoleType(), description));
+                if (newRoleVO == null) {
+                    throw new CloudRuntimeException("Unable to add role into DB, is DB full?");
```
nit - Can you improve the exception message?

sureshanaparti replied:
> nit - Can you improve the exception message?

Updated

rhtyd commented on server/src/main/java/org/apache/cloudstack/acl/RoleManagerImpl.java:
```diff
@@ -159,9 +242,6 @@ public Role updateRole(final Role role, final String name, final RoleType roleTy
             roleVO.setName(name);
         }
         if (roleType != null) {
-            if (role.getId() <= RoleType.User.getId()) {
```
Here was a check which determined that the role was built-in, i.e. the first 4 roles treated as built-in and no changes to these 4 roles were allowed by using enum idx without requiring a db column for that.

sureshanaparti replied:
@rhtyd agree that the role ids are compared against the role type enum earlier. now these roles, with ids (1,2,3,4) are marked as default roles, which is considered while updating or deleting the role. The `role` is already populated with db details.

sureshanaparti commented:
> First pass review, no major issues; mostly logistics/cosmetics/nits.
> Please ping after addressing @sureshanaparti Thanks.

@rhyd addressed the comments, please take a look. thanks.

rhtyd commented:
I've made this my top priority to conclude reviewing this. Thanks for reminding @sureshanaparti

rhtyd commented:
@sureshanaparti I've rebased to latest master.
@blueorangutan package

blueorangutan commented:
@rhtyd a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos7 ✔debian. JID-1321

rhtyd commented on engine/schema/src/main/resources/META-INF/db/schema-41400to41500.sql:
```diff
@@ -18,3 +18,81 @@
+
+-- Role permissions for Read-Only Admin
+SELECT id INTO @ReadOnlyAdminRoleId FROM `cloud`.`roles` WHERE name = 'Read-Only Admin' AND is_default = 1;
+SELECT @ReadOnlyAdminSortOrder:=-1;
```
With a debugger I found that execution fails for mvn deploydb with:
```
========> Processing upgrade: com.cloud.upgrade.DatabaseUpgradeChecker
[WARNING] 
java.lang.reflect.InvocationTargetException
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.mojo.exec.ExecJavaMojo$1.run (ExecJavaMojo.java:297)
    at java.lang.Thread.run (Thread.java:834)
Caused by: com.cloud.utils.exception.CloudRuntimeException: Unable to upgrade the database
    at com.cloud.upgrade.DatabaseUpgradeChecker.upgrade (DatabaseUpgradeChecker.java:271)
    at com.cloud.upgrade.DatabaseUpgradeChecker.check (DatabaseUpgradeChecker.java:342)
    at com.cloud.upgrade.DatabaseCreator.main (DatabaseCreator.java:217)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.mojo.exec.ExecJavaMojo$1.run (ExecJavaMojo.java:297)
    at java.lang.Thread.run (Thread.java:834)
Caused by: com.cloud.utils.exception.CloudRuntimeException: Unable to execute upgrade script
    at com.cloud.upgrade.DatabaseUpgradeChecker.runScript (DatabaseUpgradeChecker.java:209)
    at com.cloud.upgrade.DatabaseUpgradeChecker.upgrade (DatabaseUpgradeChecker.java:258)
    at com.cloud.upgrade.DatabaseUpgradeChecker.check (DatabaseUpgradeChecker.java:342)
    at com.cloud.upgrade.DatabaseCreator.main (DatabaseCreator.java:217)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.mojo.exec.ExecJavaMojo$1.run (ExecJavaMojo.java:297)
    at java.lang.Thread.run (Thread.java:834)
Caused by: java.sql.SQLException: Column index out of range.
    at com.cloud.utils.db.ScriptRunner.runScript (ScriptRunner.java:185)
    at com.cloud.utils.db.ScriptRunner.runScript (ScriptRunner.java:87)
    at com.cloud.upgrade.DatabaseUpgradeChecker.runScript (DatabaseUpgradeChecker.java:203)
    at com.cloud.upgrade.DatabaseUpgradeChecker.upgrade (DatabaseUpgradeChecker.java:258)
    at com.cloud.upgrade.DatabaseUpgradeChecker.check (DatabaseUpgradeChecker.java:342)
    at com.cloud.upgrade.DatabaseCreator.main (DatabaseCreator.java:217)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.mojo.exec.ExecJavaMojo$1.run (ExecJavaMojo.java:297)
    at java.lang.Thread.run (Thread.java:834)
```

sureshanaparti replied:
@rhtyd Moved permissions update to the Upgrade* class. Please check.

rhtyd commented:
@blueorangutan test

blueorangutan commented:
@rhtyd a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

sureshanaparti commented:
@blueorangutan package

blueorangutan commented:
@sureshanaparti a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos7 ✔debian. JID-1337

sureshanaparti commented:
@blueorangutan test

blueorangutan commented:
@sureshanaparti a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-1674)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 48018 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr4071-t1674-kvm-centos7.zip
Intermittent failure detected: /marvin/tests/smoke/test_kubernetes_clusters.py
Intermittent failure detected: /marvin/tests/smoke/test_vpc_redundant.py
Intermittent failure detected: /marvin/tests/smoke/test_vpc_vpn.py
Smoke tests completed. 82 look OK, 1 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_01_create_redundant_VPC_2tiers_4VMs_4IPs_4PF_ACL | `Failure` | 347.44 | test_vpc_redundant.py
test_03_create_redundant_VPC_1tier_2VMs_2IPs_2PF_ACL_reboot_routers | `Failure` | 342.00 | test_vpc_redundant.py

rhtyd commented:
Thanks @sureshanaparti I'll do another review pass and ping you.

rhtyd commented on engine/schema/src/main/java/com/cloud/upgrade/dao/Upgrade41400to41500.java:
```diff
@@ -235,6 +238,267 @@ private void updateSystemVmTemplates(final Connection conn) {
+        addRolePermissionsForReadOnlyAdmin(conn);
+        addRolePermissionsForReadOnlyUser(conn);
+        addRolePermissionsForAdminSupport(conn);
+        addRolePermissionsForUserSupport(conn);
```
Minor nit - can we call the role 'SupportAdmin' and 'SupportUser'?

sureshanaparti replied:
> Minor nit - can we call the role 'SupportAdmin' and 'SupportUser'?

Updated

rhtyd commented on engine/schema/src/main/java/com/cloud/upgrade/dao/Upgrade41400to41500.java:
```diff
@@ -235,6 +238,267 @@ private void updateSystemVmTemplates(final Connection conn) {
+                    }
+
+                    List<String> insertSqlForAdminSupportRolePermissions = new ArrayList<String>();
+                    insertSqlForAdminSupportRolePermissions.add("INSERT INTO `cloud`.`role_permissions` (`uuid`, `role_id`, `rule`, `permission`, `sort_order`) VALUES (UUID(), ?, 'prepareHostForMaintenance', 'ALLOW', ?) ON DUPLICATE KEY UPDATE rule=rule");
```
nit - Could be made less verbose by doing a loop on the API strings and add the string to the collection/list. Same for other parts of code.

sureshanaparti replied:
> nit - Could be made less verbose by doing a loop on the API strings and add the string to the collection/list. Same for other parts of code.

Updated with API rules

rhtyd commented on server/src/main/java/org/apache/cloudstack/acl/RoleManagerImpl.java:
```diff
@@ -146,10 +150,89 @@ public RoleVO doInTransaction(TransactionStatus status) {
     }
 
+    @Override
+    @ActionEvent(eventType = EventTypes.EVENT_ROLE_CREATE, eventDescription = "creating Role")
```
Should the description say we're creating a role by cloning another role ?

sureshanaparti replied:
> Should the description say we're creating a role by cloning another role ?

Updated event description

rhtyd commented on api/src/main/java/org/apache/cloudstack/api/command/admin/acl/ImportRoleCmd.java:
```diff
@@ -0,0 +1,139 @@
+        if (MapUtils.isEmpty(rules)) {
+            return null;
+        }
+
```
@sureshanaparti why not process rules and convert into a list of TOs/VOs, or something simpler that service layer can easily consume? Can you make the iteration logic simpler? Thanks.

sureshanaparti replied:
> @sureshanaparti why not process rules and convert into a list of TOs/VOs, or something simpler that service layer can easily consume? Can you make the iteration logic simpler? Thanks.

The Map here corresponds to a rule, with the permission details maintained in the keys: rule, permission & description, after validating them. The cmd object hold list of such maps, which are the role permissions for the role being imported.

rhtyd commented:
@sureshanaparti I've completed another pass, please see the comments. Next, I'll test/review the Primate PR against this branch.

rhtyd commented:
@blueorangutan package

blueorangutan commented:
@rhtyd a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos7 ✖debian. JID-1383

rhtyd commented:
@sureshanaparti I've converted the PR to draft, until testing concludes so we don't accidently merge this cc @borisstoyanov @vladimirpetrov
@blueorangutan package

blueorangutan commented:
@rhtyd a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos7 ✔debian. JID-1393

rhtyd commented:
@blueorangutan test

blueorangutan commented:
@rhtyd a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-1741)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 46789 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr4071-t1741-kvm-centos7.zip
Intermittent failure detected: /marvin/tests/smoke/test_kubernetes_supported_versions.py
Intermittent failure detected: /marvin/tests/smoke/test_router_dhcphosts.py
Intermittent failure detected: /marvin/tests/smoke/test_vpc_redundant.py
Smoke tests completed. 82 look OK, 1 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_01_add_delete_kubernetes_supported_version | `Error` | 1807.94 | test_kubernetes_supported_versions.py

borisstoyanov commented:
@blueorangutan package

blueorangutan commented:
@borisstoyanov a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos7 ✔debian. JID-1491

Pearl1594 commented on api/src/main/java/org/apache/cloudstack/api/command/admin/acl/UpdateRoleCmd.java:
```diff
@@ -67,13 +63,6 @@ public String getRoleName() {
-        return null;
-    }
-
     public String getRoleDescription() {
```
Since roleType and roleDescription have been moved to RoleCmd class, can we not remove getRoleDescription() from this class?

sureshanaparti replied:
> Since roleType and roleDescription have been moved to RoleCmd class, can we not remove getRoleDescription() from this class?

thanks @Pearl1594, missed this one. updated.

borisstoyanov commented:
@blueorangutan package

blueorangutan commented:
@borisstoyanov a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos7 ✔debian. JID-1496

borisstoyanov commented:
@blueorangutan package

blueorangutan commented:
@borisstoyanov a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos7 ✔debian. JID-1504

borisstoyanov commented:
@blueorangutan test

blueorangutan commented:
@borisstoyanov a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

rhtyd commented:
Cheers @borisstoyanov I'll merge this tomorrow if tests pass.

blueorangutan commented:
<b>Trillian test result (tid-1939)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 47854 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr4071-t1939-kvm-centos7.zip
Intermittent failure detected: /marvin/tests/smoke/test_password_server.py
Intermittent failure detected: /marvin/tests/smoke/test_router_dhcphosts.py
Intermittent failure detected: /marvin/tests/smoke/test_vpc_redundant.py
Intermittent failure detected: /marvin/tests/smoke/test_vpc_vpn.py
Smoke tests completed. 80 look OK, 3 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_isolate_network_password_server | `Failure` | 13.23 | test_password_server.py
test_router_dhcphosts | `Failure` | 13.45 | test_router_dhcphosts.py
ContextSuite context=TestRouterDHCPHosts>:teardown | `Error` | 25.83 | test_router_dhcphosts.py
test_03_create_redundant_VPC_1tier_2VMs_2IPs_2PF_ACL_reboot_routers | `Error` | 321.61 | test_vpc_redundant.py
test_04_rvpc_network_garbage_collector_nics | `Error` | 3870.10 | test_vpc_redundant.py

borisstoyanov commented:
@blueorangutan test

blueorangutan commented:
@borisstoyanov a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

rhtyd commented:
Test LGTM, failures not related to the PR
